### PR TITLE
TST: Add tests for ndimage converter functions.

### PR DESCRIPTION
**Pull request by jaimefrio**
This PR:
 * moves the converter functions to their own header file, `ni_converters.h`,
 * adds a Python function implemented in C, `_test_converters`,  that uses all the converters in `ni_converters.h`,
 * adds tests for the converters using this function in `test_ni_converters.py`, and
 * fixes a segfault detected by these tests due to freeing an uninitialized `PyArrayDims` pointer.

### CODE REVIEW

**Comment by rgommers** [comment]
Tests are giving ``MemeryError``s, that doesn't look quite right

**Comment by jaimefrio** [comment]
Yes, it doesn't seem quite right... But after all the subtle behavior differences we have been finding between OSX and Linux these past days, that was the whole point of writing these tests. Will take a look tonight.

**Comment by jaimefrio** [comment]
I can't reproduce any of this on OSX, so I'm going to try to figure out what's going on by adding printf's and sending the changes to Travis... Please ignore my commits until I actively request a review.

**Comment by jaimefrio** [comment]
Let's see what Travis thinks, but the `MemoryError`s should be gone...

**Comment by jaimefrio** [comment]
So I still need to fix a test where the exception raised is not the one I thought, but the other issue is fixed.

**Comment by jaimefrio** [comment]
That should be it. Unless Travis disagrees this is ready for review.

File: scipy/ndimage/tests/test_ni_converters.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
+from __future__ import division, print_function, absolute_import
+
+import gc
+import itertools
+
+import numpy as np
+from numpy.testing import (assert_, assert_array_equal, assert_raises_regex,
+                           run_module_suite)
```
Comment (pv) [inline scipy/ndimage/tests/test_ni_converters.py]: Maybe don't add run_module_suite at all, so I don't have to remove it in gh-7572

Comment (jaimefrio) [inline scipy/ndimage/tests/test_ni_converters.py]: Done!

**Comment by rgommers** [comment]
Travis disagrees. ``assert_raises_regex`` should be imported from ``scipy._lib`` not ``np.testing``. Not sure about the ``test_origin_and_new_array`` one. And the pep8 bot found something to nitpick about.

**Comment by jaimefrio** [comment]
Ah yes, PEP8 I forgot about...

And apparently there is a plethora of different errors that are raised by different combinations of Pythons and NumPys when you call `PyArray_IntpConverter` with wrong input (a string in this case), so I'm going to have to give up on the `_regex` part of `assert_raises_regex`...

Are `assert_` and `assert_array_equal` OK being imported from `np.testing` or should they also come from `scipy._lib`?

Is there a summary somewhere of how the nose to pytest transition affects the way tests are or should be written? Or should I have been paying attention to all those issues and PRs that I have been happily ignoring?

**Comment by pv** [comment]
Nose to pytest is mainly about not using run_module_suite, TestCase, or 
yield tests (superceded by pytest.mark.parametrize), and writing 
"setup_method" instead of "setup".
.
Strictly speaking you don't have to care about that before the pytest 
PRs are merged, it's just a matter of more rebasing.

**Comment by pv** [comment]
The scipy._lib._numpy_compat imports are only required for functions 
which have been introduced after Numpy 1.8.2, which is currently the 
earliest supported version. assert_raises_regex was added in numpy 1.9.0.

**Comment by jaimefrio** [comment]
Thanks for the pointers, @pv! Is my use of `setUp()` as a method in the `TestNIConverters` fair game?

Other than that I do think this is now ready...

**Comment by jaimefrio** [comment]
While this is ready, I have started working on porting the whole `nd_image.c` file to Cython. So don't worry about reviewing and merging: it will be lost like tears in the rain anyway...

**Comment by WarrenWeckesser** [comment]
> ...  it will be lost like tears in the rain anyway...

It's too bad it won't live.  But then again, who does?

**Comment by rgommers** [comment]
@jaimefrio so do you want to close this then?

**Comment by jaimefrio** [comment]
Let me push something to replace it before closing it. I have a couple of half baked PRs in the oven...

**Comment by pv** [comment]
If this is ready, please rebase on master, so we get a clean test run (I already got burned a few times by merging non-rebased stuff and then having it fail...).

**Comment by jaimefrio** [comment]
This is not to be merged, so closing it.

**Pull request closed without merging**